### PR TITLE
UI Filters

### PR DESCRIPTION
**Pull request by alex40724**
**KS Filter**

Please ignore the Toggle Button related changes. The Toggle Button has already a separte PR (https://github.com/ILIAS-eLearning/ILIAS/pull/1098).

We currently plan to wrap a service around this UI element that takes care of persistence requirements (keep values in session and so on...).

Filter rules:
https://github.com/ILIAS-eLearning/ILIAS/pull/1193/files#diff-a5b0844daf7ef3c49988ca394af4e686

Interface:
https://github.com/ILIAS-eLearning/ILIAS/pull/1193/files#diff-a4ae795be99f4117745c8d359752decc

Example:
https://github.com/ILIAS-eLearning/ILIAS/pull/1193/files#diff-9deb1a1100cc31beba9895e124efae77

Related Feature Wiki Pages
https://docu.ilias.de/goto_docu_wiki_wpage_3687_1357.html
https://docu.ilias.de/goto_docu_wiki_wpage_4303_1357.html
https://docu.ilias.de/goto_docu_wiki_wpage_3991_1357.html

I put this already on the next JF agenda at least to announce and give an overview. I understand that we might need more time to decide on this.

---
Changes for JourFixe 15 Oct 2018

**KS Elements**

FilterInput Interface:
https://github.com/ILIAS-eLearning/ILIAS/pull/1193/files#diff-89019197d3b0e8e3a515a1ad77f4e62d

Text Input Implementation of FilterInput Interface:
https://github.com/ILIAS-eLearning/ILIAS/pull/1193/files#diff-ff9ac87aef681533573690239203bad0

**Filter Service**

Example:
https://github.com/ILIAS-eLearning/ILIAS/pull/1193/files#diff-9deb1a1100cc31beba9895e124efae77

Service base factory:
https://github.com/ILIAS-eLearning/ILIAS/pull/1193/files#diff-c2f530c1407ca911dedae9983378fbbe

Sketchy Implementation of filter factory:
https://github.com/ILIAS-eLearning/ILIAS/pull/1193/files#diff-068fe6f88f68cce558ef52dc48380f14

DIC change:
https://github.com/ILIAS-eLearning/ILIAS/pull/1193/files#diff-f46de88208848386905f0e063e513b62

### CODE REVIEW

File: src/UI/Component/Input/Container/Factory.php [hunk 1]
```diff
@@ -29,4 +29,70 @@
 	 * @return  \ILIAS\UI\Component\Input\Container\Form\Factory
 	 */
 	public function form();
+
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *      Filters are used to let the user limit content within a table.
```
Comment (alex40724) [inline src/UI/Component/Input/Container/Factory.php]: table, list or any other collection of items presented on the screen.

File: src/UI/Component/Input/Container/Factory.php [hunk 2]
```diff
@@ -29,4 +29,71 @@
...
+	 *      presented on the screen.
+	 *   composition: >
+	 *      Filters are composed of two visually separated entities:
+	 *      First, there is the Filter Bar at the top. It contains an Expand/Collapse Glyph on the left side
+	 *      On the right, an "Apply" Bulky Button, a "Reset" Bulky Button and a Toggle Button are placed. An additional
+	 *      row with set filter settings will be shown at the bottom of the Filter Bar, when the Filter Bar is collapsed
+	 *      and the Filter is activated at the same time.
+	 *      Second, there is an area where so-called Filter Fields are displayed. Every Filter Field holds a Label on the left
+	 *      and a "Remove" Glyph on the right. Every Filter Field stores an Input Field. After the last Filter Field,
+	 *      an "Add" Bulky Button is always shown.
```
Comment (alex40724) [inline src/UI/Component/Input/Container/Factory.php]: "is always shown." > "is shown, if additional fields can be added."

File: src/UI/Component/Input/Container/Factory.php [hunk 3]
```diff
@@ -29,4 +29,71 @@
...
+	 *   effect: >
+	 *      In the Filter Bar:
+	 *      Clicking on the Expand/Collapse Glyph expands/collapses the second entity of the Filter,
+	 *      where the Filter Fields are placed. When the Filter Bar is expanded, the row with set filter settings
+	 *      at the bottom will be hidden.
+	 *      Clicking on the "Apply" Bulky Button applies the settings which the user has made for the Filter and
+	 *      reloads the page immediately.
+	 *      Clicking on the "Reset" Bulky Button resets the Filter to the initial state and reloads the page immediately.
```
Comment (Amstutz) [inline src/UI/Component/Input/Container/Factory.php]: Maybe change reloads the page immediatly to "resets the content of the operated container (e.g. Table) immediatly"

File: src/UI/Component/Input/Container/Factory.php [hunk 4]
```diff
@@ -29,4 +29,71 @@
...
+	 *      Clicking on the Toggle Button imitates a click on the "Apply" Bulky Button, activates/deactivates the Filter
+	 *      and reloads the page immediately.
```
Comment (Amstutz) [inline src/UI/Component/Input/Container/Factory.php]: same as above, maybe rephrase.

File: src/UI/Component/Input/Container/Filter/Factory.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *      Possible properties, which only Standard Filters have, should be mentioned here.
+	 *
+	 * rules:
+	 *   usage:
+	 *     1: Possible rules, which only Standard Filters have, should be mentioned here.
+	 *
```
Comment (Amstutz) [inline src/UI/Component/Input/Container/Filter/Factory.php:23]: I guess this will be added, note that this might be very short. Most is mentioned in the parent description.

**Review by Amstutz — COMMENTED** [review]
This comment only targets the filter desciptions, since I guess the interface is somewhat non-final yet (correct?).

@tfamula Thanks a lot for this excellent description of filters. IMO, there is not much to be added and to me this looks really like a very good examples of how such descriptions are supposed to look, thx a lot!

File: src/UI/Component/Input/Container/Filter/ProxyFilterField.php [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+/* Copyright (c) 2018 Thomas Famula <famula@leifos.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\UI\Component\Input\Container\Filter;
+
+use ILIAS\UI\Component\Component;
+use ILIAS\UI\Component\Clickable;
+use ILIAS\UI\Component\JavaScriptBindable;
+
+interface ProxyFilterField extends Component, JavaScriptBindable, Clickable {
```
Comment (Amstutz) [inline src/UI/Component/Input/Container/Filter/ProxyFilterField.php]: I did not wrap my head around this for long, why is this needed? Is this only for tagging elements? Is it to introduce the clickable behaviour to attach js? If so, maybe this is useful for more than just filters. 

Since I do not understand it imeediatly, leave a comment in the interface description.

File: src/UI/Component/Input/Container/Factory.php [hunk 7]
```diff
@@ -29,4 +29,71 @@
...
+	 *   effect: >
+	 *      In the Filter Bar:
+	 *      Clicking on the Expand/Collapse Glyph expands/collapses the second area of the Filter,
+	 *      where the Filter Fields are placed. When the Filter Bar is expanded, the row with set filter settings
+	 *      at the bottom of the Filter Bar will be hidden.
+	 *      Clicking on the "Apply" Bulky Button applies the settings which the user has made for the Filter and
+	 *      reloads the page immediately.
+	 *      Clicking on the "Reset" Bulky Button resets the Filter to the initial state and reloads the page immediately.
+	 *      Clicking on the "Toggle" Button imitates a click on the "Apply" Bulky Button, activates/deactivates the Filter
+	 *      and reloads the page immediately.
```
Comment (alex40724) [inline src/UI/Component/Input/Container/Factory.php]: When toggling "on".

**Comment by matthiaskunkel** [comment]
Jour Fixe, 24 SEP 2018 : We highly appreciate this PR and fully agree with its direction. We keep the discussion open until the next JF to allow others that couldn't participate today to add their comments.

**Comment by matthiaskunkel** [comment]
EDIT by @alex40724 : moved wrong JF comment to PR #1194

**Comment by xus** [comment]
> Jour Fixe, 24 SEP 2018 : We highly appreciate this suggestion. We decidec that icons in the snowboard should not be clickable but only the background image below and the title. We will use the outline icons offered by CaT for presenting the object type and the certificate. For the progress meter we will use the mini version plus text (and without background). We will try to solve the responsive problem of the progress meter as well.

Hi Matthias, I'm pretty sure that this comment is related to this other PR -> https://github.com/ILIAS-eLearning/ILIAS/pull/1194

File: src/UI/Component/Input/Container/Factory.php [hunk 8]
```diff
@@ -29,4 +29,71 @@
...
+
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *      Filters are used to let the user limit content within a table, list or any other collection of items
+	 *      presented on the screen.
+	 *   composition: >
+	 *      Filters are composed of two visually separated areas:
+	 *      First, there is the Filter Bar at the top. It contains an Expand/Collapse Glyph on the left side
```
Comment (klees) [inline src/UI/Component/Input/Container/Factory.php]: Very minor, but sentence suddenly ends here.

File: src/UI/Component/Input/Container/Factory.php [hunk 9]
```diff
@@ -29,4 +29,71 @@
...
+	 *      On the right, an "Apply" Bulky Button, a "Reset" Bulky Button and a Toggle Button are placed. An additional
+	 *      row with set filter settings will be shown at the bottom of the Filter Bar, when the Filter Bar is collapsed
```
Comment (klees) [inline src/UI/Component/Input/Container/Factory.php]: Does it need to be a new row? Why not place it in the middle between the glyph on the left and the button on the right?

File: src/UI/Component/Input/Container/Factory.php [hunk 10]
```diff
@@ -29,4 +29,71 @@
...
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *      Filters are used to let the user limit content within a table, list or any other collection of items
+	 *      presented on the screen.
+	 *   composition: >
+	 *      Filters are composed of two visually separated areas:
+	 *      First, there is the Filter Bar at the top. It contains an Expand/Collapse Glyph on the left side
+	 *      On the right, an "Apply" Bulky Button, a "Reset" Bulky Button and a Toggle Button are placed. An additional
```
Comment (klees) [inline src/UI/Component/Input/Container/Factory.php]: Maybe say what the Toggle Button does, this info is there for the other buttons as well.

File: src/UI/Component/Input/Container/Factory.php [hunk 11]
```diff
@@ -29,4 +29,71 @@
...
+	 *      row with set filter settings will be shown at the bottom of the Filter Bar, when the Filter Bar is collapsed
+	 *      and the Filter is activated at the same time.
+	 *      Second, there is an area where so-called Filter Fields are displayed. Every Filter Field holds a Label on the left
+	 *      and a "Remove" Glyph on the right. Every Filter Field holds an Input Field. After the last Filter Field,
+	 *      an "Add" Bulky Button is shown if additional Filter Fields can be added.
```
Comment (klees) [inline src/UI/Component/Input/Container/Factory.php]: I would propose to conceptualize the fields in the filter just as normal `Input Fields` that are styled in a certain way when used in the filter. I see that it might be necessary have a wrapper around the normal `Input Fields` in the implementation but I cannot see why we need to introduce that additional layer in the conceptual description or the code-wise interface of the filter.

Comment (klees) [reply src/UI/Component/Input/Container/Factory.php]: Also when reading further (especially the rules), this entry seems to describe two components, the `Filter` and the `Filter Field` which to me seemed odd.

Comment (klees) [reply src/UI/Component/Input/Container/Factory.php]: I think the description of the visuals is missing the value of a field in the filter.

File: src/UI/Component/Input/Container/Factory.php [hunk 12]
```diff
@@ -29,4 +29,71 @@
...
+	 *     3: The Popovers SHOULD be shown below the Filter Field or the "Add" Bulky Button.
+	 *   ordering:
+	 *     1: A Filter MUST be placed above the item collection it acts upon.
+	 *   responsiveness:
+	 *     1: >
+	 *        On screens larger than medium size, there MUST be three Filter Fields per row. On medium-sized screens
+	 *        or below, only one Filter Field MUST be shown per row.
+	 *     2: >
+	 *        On screens larger than small size, the "Apply" and "Reset" Bulky Buttons MUST contain a glyph plus text.
+	 *        On small-sized screens, these Bulky Buttons MUST contain only a glyph.
```
Comment (klees) [inline src/UI/Component/Input/Container/Factory.php]: This might actually be a rule for the `Bulky Button`.

File: src/UI/Component/Input/Container/Factory.php [hunk 13]
```diff
@@ -29,4 +29,71 @@
...
+	 *   accessibility:
+	 *     1: Filter Fields MUST be accessible by keyboard by using the "Tab"-Key and clickable by using the "Return"-Key.
```
Comment (klees) [inline src/UI/Component/Input/Container/Factory.php]: I would prefer if the fields would already be extended when someone tabs into them.

File: src/UI/Component/Input/Container/Filter/Factory.php (hunk 5 above)
Comment (klees) [reply src/UI/Component/Input/Container/Filter/Factory.php:23]: Yes, this seems to be missing. I wonder, what the other instances of filter could be. If there are none, I would remove this Factory completly. This is similar to `Form`, though, but with `Form` I see at least a `Asynchronous Form in a Modal` as second instance. I cannot see that for `Filter` atm, because we already have sufficient means to make them small.

Comment (klees) [reply src/UI/Component/Input/Container/Filter/Factory.php:23]: Ok, now I saw the comment below, that says that asynchronous data submission might be a point. Then the description for `Filter` is odd, because it explicitely says, that page is reloaded. I still tend to remove this layer.

File: src/UI/Component/Input/Container/Filter/Factory.php [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+	 * ---
+	 *
+	 * @param    string|Signal    $toggle_action_on
+	 * @param    string|Signal    $toggle_action_off
+	 * @param    string|Signal    $expand_action
+	 * @param    string|Signal    $collapse_action
+	 * @param    string|Signal    $apply_action
+	 * @param    string|Signal    $reset_action
```
Comment (klees) [inline src/UI/Component/Input/Container/Filter/Factory.php:31]: Are these actions passed here because the filter needs to know which URL to call on which action?

I don't think this is a very friendly interface. I would expect that most of the time the URLs are exactly the same and only differ by some "&cmd=toggle_on". I would still need to create all of them by hand to create a filter. I would be nicer if I could just pass one end-point and have the filter generate the different commands.

File: src/UI/Component/Input/Container/Filter/ProxyFilterField.php (hunk 6 above)
Comment (klees) [reply src/UI/Component/Input/Container/Filter/ProxyFilterField.php]: I also do not understand this.

File: src/UI/Component/Input/Container/Filter/Filter.php [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+	 * @return bool
+	 */
+	public function isExpanded();
+
+	/**
+	 * Get a Filter like this, but already expanded.
+	 *
+	 * @return Filter
+	 */
+	public function withExpanded();
```
Comment (klees) [inline src/UI/Component/Input/Container/Filter/Filter.php:114]: How would I collapse?

File: src/UI/Component/Input/Container/Filter/Filter.php [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+	 * @return bool
+	 */
+	public function isActivated();
+
+	/**
+	 * Get a Filter like this, but already activated.
+	 *
+	 * @return Filter
+	 */
+	public function withActivated();
```
Comment (klees) [inline src/UI/Component/Input/Container/Filter/Filter.php:93]: How would I deactivate?

File: src/UI/Component/Input/Container/Filter/Filter.php [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+	 * @return    array<mixed,\ILIAS\UI\Component\Input\Input>
+	 */
+	public function getInputs();
+
+	/**
+	 * Get an array with boolean values to get to know if the Inputs for the Filter are rendered initially
+	 *
+	 * @return array
+	 */
+	public function isInputRendered();
```
Comment (klees) [inline src/UI/Component/Input/Container/Filter/Filter.php]: How is this different from expanded/collapsed?

File: src/UI/Implementation/Component/Input/Field/FilterContextRenderer.php [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+
+		$tpl = $this->getTemplate("tpl.context_form.html", true, true);
+		/**
+		 * TODO: should we through an error in case for no name or render without name?
+		 *
+		 * if(!$input->getName()){
+		 * throw new \LogicException("Cannot render '".get_class($input)."' no input name given.
+		 * Is there a name source attached (is this input packed into a container attaching
+		 * a name source)?");
+		 * } */
```
Comment (klees) [inline src/UI/Implementation/Component/Input/Field/FilterContextRenderer.php]: This is a note from the standard-renderer as well, but we should not forget this. IMO the `$input` should always have a name here, but I'll check again. Thx for unintentionally bringing this up =)

**Review by klees — CHANGES_REQUESTED** [review]
Hi @alex40724,
thx a lot for this contribution.

I see room to improve the description of the filter, I hope my comments give a direction. Please feel free to ask if you need additional input (or my comments are hard to understand or anything). 

I'm definitely not happy with the programmatical interface (`$factory->input(..)`) and would not accept it like this. I might be missing something, though.

Do you need more definite change request or can you work with my other comments?

I'll also have some time on thursday or friday if you want to discuss stuff personally via skype.

Best Regards!

File: src/UI/Component/Input/Container/Factory.php (hunk 9 above)
Comment (alex40724) [reply src/UI/Component/Input/Container/Factory.php]: I would expect that this does not work well with a higher number of selected filter values (not enough space). Wrapping into new lines with a new section would be possible easier. I do not have a strong preference here. It should be a minor point to move this to another presentation and not keep us from moving forward.

File: src/UI/Component/Input/Container/Factory.php (hunk 11 above)
Comment (alex40724) [reply src/UI/Component/Input/Container/Factory.php]: The lables and remove glyphs are not part of the **Input Fields**. They are attached to "something" that wraps the input fields and occur multiple times within the **Filter**. So the lables and remove glyphs are part of something smaller than the Filter which is not an Input Field. For us it was easier if we give this "something" a name. Otherwise it is hard to address this concept in the description and rules.

Can you suggest an alternative text?

.

File: src/UI/Component/Input/Container/Filter/Factory.php (hunk 14 above)
Comment (alex40724) [reply src/UI/Component/Input/Container/Filter/Factory.php:31]: The action concept is currently the "atomic" concept for "commands" in the UI framework in my understanding. Do we want the UI framework to get knowledge and partial control on URL/request parameters? We can do this, but I though this would become messy.

Our idea is that we need an additional Filter service anyway that keeps track of selected values and persistence. This Filter service will offer a simplified API for the actions as well. Mostly consumers will only use this Filter service.

File: src/UI/Component/Input/Container/Filter/Filter.php (hunk 17 above)
Comment (alex40724) [reply src/UI/Component/Input/Container/Filter/Filter.php]: Expand/collapse is a property of the Filter. Being rendered or not is a property of a Filter Field. But I am not sure if this array helps the consumer in any way, too. I will discuss this with @tfamula. Maybe we just remove this function.

File: src/UI/Component/Input/Container/Filter/Filter.php (hunk 16 above)
Comment (alex40724) [reply src/UI/Component/Input/Container/Filter/Filter.php:93]: @tfamula: Yes, this seems to miss a bool parameter.

File: src/UI/Component/Input/Container/Filter/Filter.php (hunk 15 above)
Comment (alex40724) [reply src/UI/Component/Input/Container/Filter/Filter.php:114]: @tfamula: Again a missing bool parameter.

File: src/UI/Component/Input/Container/Filter/ProxyFilterField.php (hunk 6 above)
Comment (alex40724) [reply src/UI/Component/Input/Container/Filter/ProxyFilterField.php]: This is an internal class for the Filter Fields. It will not be offered through the factory. We can also remove it from the interfaces, if needed. This internal class wants to reuse framework concepts like JSBindable and Clickable, but I am pretty sure that this class is not useful for other components. It's very focused on the **Filter Field** internals (e.g. opening the popovers "on click"). We will add some comments.

File: src/UI/Implementation/Component/Input/Field/FilterContextRenderer.php [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+		$content = $this->renderInputFieldWithContext($input_tpl, $input);
+		$popover = $f->popover()->standard($f->legacy($content))->withVerticalPosition();
+		$tpl->setVariable("POPOVER", $default_renderer->render($popover));
+
+		$prox = new ProxyFilterField();
+		$prox = $prox->withOnClick($popover->getShowSignal());
+
+		$this->maybeRenderId($prox, $tpl);
+		return $tpl->get();
+	}
```
Comment (alex40724) [inline src/UI/Implementation/Component/Input/Field/FilterContextRenderer.php:149]: Here is the proxy field again... As you can see it internally attaches the popovers with the Input Fields to the Filter Fields.

File: src/UI/Component/Input/Container/Factory.php (hunk 13 above)
Comment (alex40724) [reply src/UI/Component/Input/Container/Factory.php]: This is what we would like to achieve, too. Not sure how complicated this will be in the implementation, since the wrapped Input Fields will have internal tab handling, too. This might become a tricky part.

File: src/UI/Component/Input/Container/Filter/Factory.php (hunk 5 above)
Comment (alex40724) [reply src/UI/Component/Input/Container/Filter/Factory.php:23]: Ok, I though it would be "good practice" to have this layer for future extensibility. Even if we do not have an idea for future versions of Filters yet.

We currently see this being a problem with the Cards. We didn't have cards()->standard(). This forces us now, when introducing a second type of Card, to refactor all consumer code that used Cards before. Btw. dropdown->standard() is also an accepted thing. Not sure if we had any ideas of other future Dropdowns versions back then.

Comment (klees) [reply src/UI/Component/Input/Container/Filter/Factory.php:23]: Yeah, this is a valid point as well. My opinion to remove this is not very strong. What does @Amstutz say?

File: src/UI/Component/Input/Container/Filter/ProxyFilterField.php (hunk 6 above)
Comment (klees) [reply src/UI/Component/Input/Container/Filter/ProxyFilterField.php]: IMO this could just go to `Implementation`, then.

File: src/UI/Component/Input/Container/Factory.php (hunk 11 above)
Comment (klees) [reply src/UI/Component/Input/Container/Factory.php]: Something like: 
```
Second, there is an area where the Input Fields in the filter are
displayed. The fields are rendered with their label on the left and
a "Remove"-Glyph on the right.  the values, that are currently
entered in the Input fields, are displayed as non-editable text
and will get editable when the user focuses an Input Field
(see effect).
```
The difference of this is not huge, but talking about how `Input Fields` are rendered in the `Filter` will spare us a term (and hence term, definition, concept, ...).

File: src/UI/Component/Input/Container/Filter/Factory.php (hunk 14 above)
Comment (klees) [reply src/UI/Component/Input/Container/Filter/Factory.php:31]: Ok, sounds reasonable for the moment. I would expect that the UI-Framework needs to get knowledge on how links are created at some point, but this is not that point.

File: src/UI/Component/Input/Container/Filter/Factory.php (hunk 5 above)
Comment (Amstutz) [reply src/UI/Component/Input/Container/Filter/Factory.php:23]: I think if there is a valid possibility that there might be future members of the family, we should introduce the standard. To me this is the case with dropdowns (e.g. with Icons, or some very restricted usage with a some special layout etc.). It is definitely the case with cards and I think it is also the case with filters.

File: src/UI/Component/Input/Container/Factory.php (hunk 10 above)
Comment (matthiaskunkel) [reply src/UI/Component/Input/Container/Factory.php]: Jour Fixe, 08 OCT 2018 : Yes, we agree. Sentence for Toggle button should be added.

File: src/UI/Component/Input/Container/Factory.php (hunk 11 above)
Comment (matthiaskunkel) [reply src/UI/Component/Input/Container/Factory.php]: Jour Fixe, 08 OCT 2018 : We would like simply to speak of "Input Fields" (that are rendered in a specific way).

File: src/UI/Component/Input/Container/Factory.php (hunk 12 above)
Comment (matthiaskunkel) [reply src/UI/Component/Input/Container/Factory.php]: Jour Fixe, 08 OCT 2018 : Please remove " On small-sized screens, these Bulky Buttons MUST contain only a glyph." and add it to the rules of the Bulky Button.

File: src/UI/Component/Input/Container/Filter/Filter.php (hunk 15 above)
Comment (matthiaskunkel) [reply src/UI/Component/Input/Container/Filter/Filter.php:114]: Jour Fixe, 08 OCT 2018 : We agree!

File: src/UI/Component/Input/Container/Filter/Filter.php (hunk 16 above)
Comment (matthiaskunkel) [reply src/UI/Component/Input/Container/Filter/Filter.php:93]: Jour Fixe, 08 OCT 2018 : We agree!

File: src/UI/Component/Input/Container/Filter/Filter.php (hunk 17 above)
Comment (matthiaskunkel) [reply src/UI/Component/Input/Container/Filter/Filter.php]: Jour Fixe, 08 OCT 2018 : We currently prefer to remove this method.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 08 OCT 2018 : We highly appreciate this UI service and accept it for 5.4 with the recommended changes.

File: src/UI/Component/Input/Container/Filter/Factory.php (hunk 5 above)
Comment (tfamula) [reply src/UI/Component/Input/Container/Filter/Factory.php:23]: Since we just have a Standard Filter at the moment and possible differences from future Filter member are unknown yet, I would leave the description and rules empty for now.

File: src/UI/Component/Input/Container/Factory.php (hunk 7 above)
Comment (tfamula) [reply src/UI/Component/Input/Container/Factory.php]: The described function of the Toggle Button goes for both ways, toggling "on" and toggling "off".

File: src/UI/Component/Input/Container/Filter/Factory.php (hunk 5 above)
Comment (Amstutz) [reply src/UI/Component/Input/Container/Filter/Factory.php:23]: No don't do that. Check the other examples, where there is just one standard field. Obviously you do not need lots of text, but please do not leave the description empty.

File: src/UI/Component/Input/Container/Factory.php (hunk 10 above)
Comment (tfamula) [reply src/UI/Component/Input/Container/Factory.php]: Like for the other buttons, this info can be found in the section "effect": 
"Clicking on the 'Toggle' Button imitates a click on the 'Apply' Bulky Button, activates/deactivates the Filter..."

Or do you mean a little subordinate clause should be added here, too?

File: src/UI/Component/Input/Container/Factory.php (hunk 12 above)
Comment (tfamula) [reply src/UI/Component/Input/Container/Factory.php]: Should the new rule for the Bulky Button be part of this PR?

Comment (klees) [reply src/UI/Component/Input/Container/Factory.php]: IMO yes.

File: src/UI/Component/Input/Container/Factory.php (hunk 10 above)
Comment (klees) [reply src/UI/Component/Input/Container/Factory.php]: Yes. This would make this more comprehensible IMO.

**Comment by alex40724** [comment]
@Amstutz @klees : We outlined the basics of the filter service in the lower part of the description of this pull request (see top of this page).

- Is the location ok? Services/UI
- Is the naming ok? $DIC->uiService()->...
- The signature of the filter factory is very similar to the KS element. We simplified the action parameter. We also need a filter ID, since we need a key for persisting the data.

**Comment by klees** [comment]
Hi everybody!

I'm fine with that `ilUIService` **for now** but I think we need to have a discussion whether we really want to adopt this architecture for the future. I think we had a similar discussion (what to do with UI-related persistence) when these "emotion"-buttons were proposed for the timeline and I see that we will have even more similar discussions in the future for other UI-elements. I understand that there is too little time left for that discussion until the beta for 5.4 needs to be ready, but I do think we should make a conscious decision here, thus discuss this topic for 6.0 and not take this pattern for granted just now.

I had a shallow discussion with @Amstutz regarding implications and alternatives. There might be good reasons to implement the architecture proposed here, but IMO we need to weigh pros and cons before we finally decide. Regarding implications, I see the following questions and problems that we should consider:

* We intend the UI-framework to be a catalogue and documentation for all UI-elements in ILIAS at some point in the future. What are the implications here if we have a second component that takes care of UI-stuff? How is that component documented? Should we connect the documentations of the two components? How?
* For the UI-framework we have established processes and patterns to implement and change elements. Do we need these processes and patterns for a second component? How do they look like? How are they related to the UI-framework?
* For this PR I agreed for the moment to live with an interface that to me looks error prone. Do we consider "This will be used through the ilUIService most of the time." to be a valid argument to have unfriendly or error prone interfaces for consumers of the UI-framework in the future? Is the assessment, that most filters (or other future components) will mostly be used through this new component really correct?
* And, maybe the root of all the above questions: How does the general architecture of the second component look like? How are responsibilities distributed between the two components?

I think there are alternatives to the currently proposed architecture along the lines of these responsibilities, where the two extremes might be "ui framework only provides rendering" (like ilTemplate) or "ui framework knows everything about persistence and maybe other stuff" (not yet implement anywhere afaik). It would indeed be possible, e.g., to have an "ilUIPersistenceService", that implements persistence interfaces defined by the UI-framework in different ways (over `$ilDB`, over the users session, over get-parameters, ...). I'm sure there also are other valid distributions of responsibilities or maybe even more additional services we might need.

Let's have a careful look into these questions and problems once the dust around 5.4 has settled.

Best regards!

**Comment by alex40724** [comment]
I agree with all of this. The evil in this case is the $_SESSION dependency. If the UI Framwork would accept a key/value store like interface in its top construtor (injected in the DIC initialisation) much more of the implementation could stay within the framework (it could "ask" this object to keep the values).

To make the framework still usable "out of the box", it could provide standard implementations for these dependency wrappers. This would also allow to make the injection optional in the constructor.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 15 OCT 2018 : We highly appreciate the changes suggested by Alexander and accept them for 5.4. We will pick-up the discussion again after 5.4 for the future of the UI service.

File: src/UI/templates/js/Popover/popover.js [hunk 20]
```diff
@@ -93,9 +93,22 @@ il.UI = il.UI || {};
...
+
+            // webui is moving the content at the end of the document which makes it impossible to use
+            // popovers in forms without specifying a container here. We search for upper il-popover-container.
+            // If given we use this element as a container for the popover
+            var container;
+			console.log(options);
+            if (container = $('#' + $triggerer.attr('id')).parents(".il-popover-container")[0]) {
+				options = $.extend({}, {container: container}, options);
+            }
+			console.log(options);
```
Comment (Amstutz) [inline src/UI/templates/js/Popover/popover.js:105]: Sure you want them here?

File: src/UI/templates/js/Popover/popover.js [hunk 21]
```diff
@@ -93,9 +93,22 @@ il.UI = il.UI || {};
             if (WebuiPopovers.isCreated('#' + $triggerer.attr('id'))) {
                 return true;
             }
+
+            // webui is moving the content at the end of the document which makes it impossible to use
+            // popovers in forms without specifying a container here. We search for upper il-popover-container.
+            // If given we use this element as a container for the popover
+            var container;
+			console.log(options);
```
Comment (Amstutz) [inline src/UI/templates/js/Popover/popover.js:101]: Sure you want this here?

**Comment by Amstutz** [comment]
To avoid any confusion, Ill close this down. The most recent one is: #1735.
